dvc/repo/plot.py

```diff
@@ -0,0 +1,208 @@
+import logging
+import os
+from collections import OrderedDict
+
+from funcy import first
+from ruamel import yaml
+
+from dvc.exceptions import DvcException
+from dvc.plot import Template
+from dvc.repo import locked
+
+logger = logging.getLogger(__name__)
+
+
+class NoMetricInHistoryError(DvcException):
+    def __init__(self, path, revisions):
+        super().__init__(
+            "Could not find '{}' on any of the revisions: "
+            "'{}'".format(path, ", ".join(revisions))
+        )
+
+
+class NoMetricOnRevisionError(DvcException):
+    def __init__(self, path, revision):
+        self.path = path
+        self.revision = revision
+        super().__init__(
+            "Could not find '{}' on revision: " "'{}'".format(path, revision)
+        )
+
+
+class TooManyDataSourcesError(DvcException):
+    def __init__(self, datafile, template_datafiles):
+        super().__init__(
+            "Unable to reason which of possible data sources: '{}' "
+            "should be replaced with '{}'".format(
+                ", ".join(template_datafiles), datafile
+            )
+        )
+
+
+class NoDataNorTemplateProvided(DvcException):
+    def __init__(self):
+        super().__init__("Datafile or template is not specified.")
+
+
+class PlotMetricTypeError(DvcException):
+    def __init__(self, path):
+        super().__init__(
+            "'{}' - file type error\n"
+            "Only json, yaml, csv and tsv types are supported.".format(path)
+        )
+
+
+WORKSPACE_REVISION_NAME = "workspace"
+
+
+def _parse(datafile, default_plot, tree, loading_function):
+    with tree.open(datafile, "r") as fobj:
+        data = loading_function(fobj)
+        assert isinstance(data, list)
+    if default_plot:
+        assert all(len(e) >= 1 for e in data)
+        last_key = list(first(data).keys())[-1]
+        data = [{"y": d[last_key], "x": i} for i, d in enumerate(data)]
+    return data
+
+
+def _parse_yaml(datafile, default_plot, tree):
+    def load_yaml(fobj):
+        return yaml.load(fobj)
+
+    return _parse(datafile, default_plot, tree, load_yaml)
+
+
+def _parse_json(datafile, default_plot, tree):
+    def load_json(fobj):
+        return json.load(fobj, object_pairs_hook=OrderedDict)
+
+    return _parse(datafile, default_plot, tree, load_json)
+
+
+def _parse_csv(datafile, default_plot, tree, delimiter=","):
+    with tree.open(datafile, "r") as fobj:
+        if default_plot:
+            data = []
+            for index, row in enumerate(csv.reader(fobj, delimiter=delimiter)):
+                assert len(row) >= 1
+                if index == 0 and len(row) > 1:
+                    # skip header
+                    continue
+                data.append({"y": row[-1], "x": index})
+        else:
+            data = [
+                row
+                for row in (
+                    csv.DictReader(
+                        fobj, skipinitialspace=True, delimiter=delimiter
+                    )
+                )
```

Comment:
This is really hard to read. Just use some temporary vars, they are free :slightly_smiling_face: 
![image](https://user-images.githubusercontent.com/5367102/79256000-f4296c80-7e8f-11ea-89cc-f0697612975c.png)